basedtyping/__init__.py

```diff
@@ -0,0 +1,201 @@
+"""the main ``basedtyping`` module. the types/functions defined here can be used at both type-time and at runtime."""
+from __future__ import annotations
+
+from types import UnionType
+from typing import (
+    TYPE_CHECKING,
+    Callable,
+    Generic,
+    NoReturn,
+    Protocol,
+    Sequence,
+    TypeVar,
+    _GenericAlias,
+    cast,
+)
+
+from basedtyping.runtime_only import OldUnionType
+
+if TYPE_CHECKING:
+    Function = Callable[..., object]  # type:ignore[misc]
+    """any ``Callable``. useful when using mypy with ``disallow-any-explicit`` due to https://github.com/python/mypy/issues/9496
+
+    cannot actually be called unless it's narrowed, so it should only really be used as a bound in a ``TypeVar``"""
+else:
+    # for isinstance checks
+    Function = Callable
+
+# unlike the generics in other modules, these ones are meant to be imported to save you from the boilerplate
+
+T = TypeVar("T")
+T_co = TypeVar("T_co", covariant=True)
+T_cont = TypeVar("T_cont", contravariant=True)
+Fn = TypeVar("Fn", bound=Function)
+
+
+class _ReifiedGenericAlias(_GenericAlias, _root=True):
+    def __call__(self, *args: NoReturn, **kwargs: NoReturn) -> object:
+        """copied from `_GenericAlias.__call__` but modified to call `_actual_call` instead of `__call__`,
+        and throw an error if there are any unbound generics"""
+        if not self._inst:
+            raise TypeError(
+                f"Type {self._name} cannot be instantiated; "
+                f"use {self.__origin__.__name__}() instead"
+            )
+        self._check_generics_reified()
+        result = cast(_ReifiedGenericMetaclass, self.__origin__)._actual_call(
+            *args, **kwargs
+        )
+        try:
+            result.__orig_class__ = self  # type:ignore[attr-defined]
+        except AttributeError:
+            pass
```

Comment:
`__orig_class__` is basically a necessity right? I think we should fail if we can't set it.